flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,177 @@
+
+		if (enableMultiThreadDownload(restoringThreadNum)) {
+			runTransferWithMultipleThreads(runnables, restoringThreadNum, closeableRegistry);
+		} else {
+			runTransferWithSingleThread(runnables);
+		}
+	}
+
+	private static List<Runnable> createDownloadRunnables(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		CloseableRegistry closeableRegistry
+		) {
+		List<Runnable> runnables = new ArrayList<>(stateHandleMap.size());
+		for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
+			StateHandleID stateHandleID = entry.getKey();
+			StreamStateHandle remoteFileHandle = entry.getValue();
+
+			Path path = new Path(restoreInstancePath, stateHandleID.toString());
+			runnables.add(() -> downloadDataForStateHandle(path, remoteFileHandle, closeableRegistry));
+		}
+		return runnables;
+	}
+
+	private static void runTransferWithMultipleThreads(
+		List<Runnable> runnables,
+		int threadNum,
+		CloseableRegistry closeableRegistry) throws ExecutionException, InterruptedException, IOException {
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(threadNum);
+		Closeable shutDownAllTasksAndExecutorService = () -> {
+			if (executorService != null) {
+				executorService.shutdownNow();
+			}
+		};
+		closeableRegistry.registerCloseable(shutDownAllTasksAndExecutorService);
+
+		try {
+			List<CompletableFuture<Void>> futures = new ArrayList<>(runnables.size());
+			for (Runnable runnable : runnables) {
+				CompletableFuture<Void> future = CompletableFuture.runAsync(runnable, executorService);
+				futures.add(future);
+			}
+			FutureUtils.waitForAll(futures).get();
+		} finally {
+			if (closeableRegistry.unregisterCloseable(shutDownAllTasksAndExecutorService)) {
+				shutDownAllTasksAndExecutorService.close();
+			}
+		}
+	}
+
+	private static void runTransferWithSingleThread(List<Runnable> runnables) {
+		for (Runnable runnable : runnables) {
+			runnable.run();
+		}
+	}
+
+	/**
+	 * Copies the file from a single state handle to the given path.
+	 */
+	private static void downloadDataForStateHandle(
+		Path restoreFilePath,
+		StreamStateHandle remoteFileHandle,
+		CloseableRegistry closeableRegistry) {
+
+		FSDataInputStream inputStream = null;
+		FSDataOutputStream outputStream = null;
+
+		try {
+			FileSystem restoreFileSystem = restoreFilePath.getFileSystem();
+			inputStream = remoteFileHandle.openInputStream();
+			closeableRegistry.registerCloseable(inputStream);
+
+			outputStream = restoreFileSystem.create(restoreFilePath, FileSystem.WriteMode.OVERWRITE);
+			closeableRegistry.registerCloseable(outputStream);
+
+			byte[] buffer = new byte[8 * 1024];
+			while (true) {
+				int numBytes = inputStream.read(buffer);
+				if (numBytes == -1) {
+					break;
+				}
+
+				outputStream.write(buffer, 0, numBytes);
+			}
+		} catch (IOException e) {
+			throw new RuntimeException("Copy state from DFS error.", e);
```

Comment:
Throwing `RuntimeExceptions` is discouraged. Please add the `throws IOException` clause and use a `ThrowingRunnable.unchecked` when giving it to the `Executor`.